### DISCUSSION TITLE
Preferred directory structure for building a pocketbase based application.

### DISCUSSION THREAD

**Discussion by mukhtharcm** [pocketbase/pocketbase#5184] [Q&A]
I'd like to create an app like Presentator. ie, extend PocketBase with go to add a few more routes, a front with SvelteKit.
The main thing is, I want the the front end to be embedded in the go binary.
Is the Presentator directory structure the preferred way to create it?
ie, with the backend files in the root and the sk front end in UI folder?.

also, thank you very much for creating such a library :)

**Comment by ganigeorgiev** [comment]
The project structure in [`presentator/presentator`](https://github.com/presentator/presentator) is just my personal preference. Presentator doesn't need too much additional Go code (_most of the Go code are tests_) and as such it doesn't require for example creating subpackages and a mostly flat project structure works fine in this specific case.

With that said, there is no fixed structure, nor hard recommendation, how to organize your project as it will depend on the context of your actual code. As a general advice, don't blindly follow some convention and try to keep it as simple as possible. During actual development and the use cases that will arise from it, the process itself will guide/nudge you into the suitable structure for your project.

About the static files embedding - it relies on the [`//go:embed`](https://pkg.go.dev/embed) directive and there is nothing really PocketBase specific here. If you are going to use SvelteKit (not plain Svelte) with the [static adapter](https://kit.svelte.dev/docs/adapter-static) or the `ssr = false` option, note that [hash-based routing is not supported](https://github.com/sveltejs/kit/issues/7443) and when embedding the static files you need to set the `indexFallback` argument of `StaticDirectoryHandler` to `true` in order the urls of your static app to work (ex. `apis.StaticDirectoryHandler(yourFS, true)`).

**Reply by mukhtharcm** [reply]
thanks for the fast reply!
I've used Pocketbase for other projects. The way I did this was by creating a backend folder in the root of my sveltekit project and putting the custom pocketbase code there.
But that was without embedding the front end in it.
But now that I'm planning to embed the front end with the binary, that's why I asked for a recommended approach.
I think I'll go with my current setup of putting backend in the root of the sveltekit folder :)

Also, thanks' for the heads up about sveltekit. I'll report back if it worked :)

**Reply by mukhtharcm** [reply]
This approach isn't working. becasue, we can use `../build` in the embed directive.

Let me check if something can be done :)

**Comment by mukhtharcm** [comment]
I think I got it working.

Everything is in the root folder now :) and this is my pocketbase extending code.

```go
package main

import (
	"embed"
	"log"
	"os"

	"github.com/labstack/echo/v5"
	"github.com/labstack/echo/v5/middleware"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
)

//go:embed all:build
var distDir embed.FS

// DistDirFS contains the embedded dist directory files (without the "dist" prefix)
var appBuildDirFs = echo.MustSubFS(distDir, "build")

func main() {
	app := pocketbase.New()

	// serves static files from the provided public dir (if exists)
	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.GET("/*", apis.StaticDirectoryHandler(os.DirFS("./pb_public"), false))

		e.Router.GET(
			"/*",
			apis.StaticDirectoryHandler(appBuildDirFs, true),
			// uiCacheControl(),
			middleware.Gzip(),
		)

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

Will there be any problem if the app is being served from the root? ie from `localhost:8090` itself?
I'm asking this because, I've a dynamic root at my app at `/[id]/`.

Which will take precedence when a pocketbase route and svetekit route are same?

ie there's a custom pocketbase route at /blablah and a sveltekit route at /blablah.

What will happen?

**Reply by mukhtharcm** [reply]
Sorry again!
I think I'll be safe If i adhere to this advice
<img width="872" alt="image" src="https://github.com/pocketbase/pocketbase/assets/56378562/19dd4971-fb69-48da-9001-cbfa52db2686">
from this page [extending with go - routing](https://pocketbase.io/docs/go-routing/)

**Comment by mukhtharcm** [comment]
Also, while setting up I noticed that by default, while indexFallback is true, it redirects to the default `index.html` file.

I had my fallback page in svelteKit as another page than the index.html.

I've copied and slightly modified the `StaticDirectoryHandler` to accept optional fallback page.
```go
func StaticDirectoryHandler(fileSystem fs.FS, indexFallback bool, fallbackPage ...string) echo.HandlerFunc {
	return func(c echo.Context) error {
		p := c.PathParam("*")

		// escape url path
		tmpPath, err := url.PathUnescape(p)
		if err != nil {
			return fmt.Errorf("failed to unescape path variable: %w", err)
		}
		p = tmpPath

		// fs.FS.Open() already assumes that file names are relative to FS root path and considers name with prefix `/` as invalid
		name := filepath.ToSlash(filepath.Clean(strings.TrimPrefix(p, "/")))

		fileErr := c.FileFS(name, fileSystem)

		// Determine the fallback page
		fallback := "index.html"
		if len(fallbackPage) > 0 {
			fallback = fallbackPage[0]
		}

		if fileErr != nil && indexFallback && errors.Is(fileErr, echo.ErrNotFound) {
			return c.FileFS(fallback, fileSystem)
		}

		return fileErr
	}
}
```

If you like, I can open a pull request with this change  :)

**Reply by mukhtharcm** [reply]
Also just wanted to mention that, because of this pocketbase library, I've started to learn golang. With a little bit of help from GPT :)

**Reply by ganigeorgiev** [reply]
Based on your previous replies I'm too confused to understand what you are doing anymore but there are no plans to change the current behavior.

As a last attempt to clarify some eventual misunderstanding, note that the "fallback" in the above is supposed to mean that every nonexisting url matching your static route will resolve to the root/start page of your app so that it can handle the pretty url navigation (and for most of the cases you want the fallback page to be `index.html` so that users don't have to type it in the url).

**Reply by mukhtharcm** [reply]
I'll try to clarify.
According to sveltekit docs [here](https://kit.svelte.dev/docs/single-page-apps) 
>The fallback page is an HTML page created by SvelteKit from your page template (e.g. app.html) that loads your app and navigates to the correct route. For example Surge, a static web host, lets you add a 200.html file that will handle any requests that don't correspond to static assets or prerendered pages.

So in the `StaticDirectoryHandler` function, it is set as `index.html` (or so assumed from the behavior and after inspecting the code).

When I set `index.html` as fallback in my `svelte.config.js`, my homepage of the sveltekit site wasn't getting pre-rendered. ie, I was getting following warning.
> > Using @sveltejs/adapter-static
Overwriting build/index.html with fallback page. Consider using a different name for the fallback.

ie, it was rendered on browser. 
I wanted to use the home page as a landing page, so browser rendering a landing page is bad as I've heard.

So I set the fallback page to `404.html` in `svelte.config.js` and as I've told on the previous comment, I copied the `StaticDirectoryHandler` function and modified to set the fallback as `404.html`

I think I've clarified as per my understanding. If I'm wrong, please correct me.
